Issue with Rcpp and Intel compiler exposed by building haven

When installing haven 1.0.0 with the Intel compiler I endup getting this error in Rcpp files. Hadley suggested this is an Rcpp issue (I reported this for haven already):

Installing haven 1.0.0 with the Intel compiler (icpc (ICC) 15.0.3 20150407) I get this error log:

```
icpc -I/CHBS/apps/R/3.2.3/lib64/R/include        -I/usr/local/include -I"/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include" -I"/CHBS/apps/dev_apps/R/dev_lib/3.2/BH/include"  -Ireadstat -fpic  -march=native -O3  -DEIGEN_USE_MKL_ALL -DEIGEN_NO_DEBUG -DNDEBUG -fp-model precise -fp-model source -mkl=sequential                                                           -c DfReader.cpp -o DfReader.o
In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(86),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(153): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR double ZERO() { return 0.0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(86),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(154): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR double ONE() { return 1.0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(86),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(162): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR int ZERO() { return 0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(86),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(163): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR int ONE() { return 1; }
                        ^

DfReader.cpp(355): error: more than one operator "==" matches these operands:
            built-in operator "pointer == pointer"
            function "Rcpp::operator==(Rcpp::Na_Proxy, SEXP)"
            operand types are: Rcpp::AttributeProxyPolicy<Rcpp::RObject_Impl<Rcpp::PreserveStorage>>::AttributeProxy == SEXP
          if (col.attr("class") == R_NilValue) {
                                ^

compilation aborted for DfReader.cpp (code 2)
make: *** [DfReader.o] Error 2
ERROR: compilation failed for package ‘haven’
* removing ‘/CHBS/apps/dev_apps/R/dev_lib/3.2/haven’

The downloaded source packages are in
	‘/export/scratch/SGE/2415135.1.interactive.q/RtmpfjPEjz/downloaded_packages’
Warning message:
In install.packages("haven") :
  installation of package ‘haven’ had non-zero exit status

```

Here is my session info:

```
R3.2.3> sessionInfo()
R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server release 6.7 (Santiago)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] tools_3.2.3

```

Could it be that some C++ being used is not fully standard?

Thanks so much for the effort. I tried as you suggested, but it did not work. I then did a few tests in small tests as follows:

compilation with vanilla Rcpp/alogrithm.hpp and turned C++11 on:

```
c++11 compile:

icpc -I/CHBS/apps/R/3.2.3/lib64/R/include        -I/usr/local/include -I"/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include" -I"/home/weberse2/R/x86_64-pc-linux-gnu-library/3.2new/BH/include"  -Ireadstat -fpic  -std=c++11 -march=native -O3  -DEIGEN_USE_MKL_ALL -DEIGEN_NO_DEBUG -DNDEBUG -fp-model precise -fp-model source -mkl=sequential                                                           -c DfReader.cpp -o DfReader.o
In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(153): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR double ZERO() { return 0.0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(154): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR double ONE() { return 1.0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(162): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR int ZERO() { return 0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(163): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR int ONE() { return 1; }
                        ^

DfReader.cpp(355): error: more than one operator "==" matches these operands:
            built-in operator "pointer == pointer"
            function "Rcpp::operator==(Rcpp::Na_Proxy, SEXP)"
            operand types are: Rcpp::AttributeProxyPolicy<Rcpp::RObject_Impl<Rcpp::PreserveStorage>>::AttributeProxy == SEXP
          if (col.attr("class") == R_NilValue) {
                                ^

compilation aborted for DfReader.cpp (code 2)
make: *** [DfReader.o] Error 2
ERROR: compilation failed for package ‘haven’
```

So it looks to me as if the C++11 detection which you do does not work for Intel. OK, so I thought lets force the constexpr definition (by just defining as constexpr without the ifdef) and run with C++11 again:

```
icpc -I/CHBS/apps/R/3.2.3/lib64/R/include        -I/usr/local/include -I"/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include" -I"/home/weberse2/R/x86_64-pc-linux-gnu-library/3.2new/BH/include"  -Ireadstat -fpic  -std=c++11 -march=native -O3  -DEIGEN_USE_MKL_ALL -DEIGEN_NO_DEBUG -DNDEBUG -fp-model precise -fp-model source -mkl=sequential                                                           -c DfReader.cpp -o DfReader.o
DfReader.cpp(355): error: more than one operator "==" matches these operands:
            built-in operator "pointer == pointer"
            function "Rcpp::operator==(Rcpp::Na_Proxy, SEXP)"
            operand types are: Rcpp::AttributeProxyPolicy<Rcpp::RObject_Impl<Rcpp::PreserveStorage>>::AttributeProxy == SEXP
          if (col.attr("class") == R_NilValue) {
                                ^

compilation aborted for DfReader.cpp (code 2)
make: *** [DfReader.o] Error 2
ERROR: compilation failed for package ‘haven’
```

Now the constexpr is used, but there is still an error.

To complete things I furthermore compiled with no c++11 and forced definition with const. Then I get

```
icpc -I/CHBS/apps/R/3.2.3/lib64/R/include        -I/usr/local/include -I"/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include" -I"/home/weberse2/R/x86_64-pc-linux-gnu-library/3.2new/BH/include"  -Ireadstat -fpic  -march=native -O3  -DEIGEN_USE_MKL_ALL -DEIGEN_NO_DEBUG -DNDEBUG -fp-model precise -fp-model source -mkl=sequential                                                           -c DfReader.cpp -o DfReader.o
In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(154): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR double ZERO() { return 0.0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(155): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR double ONE() { return 1.0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(163): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR int ZERO() { return 0; }
                        ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(164): warning #858: type qualifier on return type is meaningless
          static inline RCPP_CONSTEXPR int ONE() { return 1; }
                        ^

DfReader.cpp(355): error: more than one operator "==" matches these operands:
            built-in operator "pointer == pointer"
            function "Rcpp::operator==(Rcpp::Na_Proxy, SEXP)"
            operand types are: Rcpp::AttributeProxyPolicy<Rcpp::RObject_Impl<Rcpp::PreserveStorage>>::AttributeProxy == SEXP
          if (col.attr("class") == R_NilValue) {
                                ^
```

Finally, I compile again without C++11, but forcing as you suggest RCPP_CONSTEXPR to be empty:

```
icpc -I/CHBS/apps/R/3.2.3/lib64/R/include        -I/usr/local/include -I"/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include" -I"/home/weberse2/R/x86_64-pc-linux-gnu-library/3.2new/BH/include"  -Ireadstat -fpic  -march=native -O3  -DEIGEN_USE_MKL_ALL -DEIGEN_NO_DEBUG -DNDEBUG -fp-model precise -fp-model source -mkl=sequential                                                           -c DfReader.cpp -o DfReader.o
In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(152): error: a member with an in-class initializer must be const
          static RCPP_CONSTEXPR int RTYPE = REALSXP;
                                          ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(161): error: a member with an in-class initializer must be const
          static RCPP_CONSTEXPR int RTYPE = INTSXP;
                                          ^

In file included from /CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp.h(94),
                 from DfReader.cpp(1):
/CHBS/apps/dev_apps/R/dev_lib/3.2/Rcpp/include/Rcpp/algorithm.h(171): error: a member with an in-class initializer must be const
          static RCPP_CONSTEXPR int RTYPE = helper_type::RTYPE;
                                          ^

DfReader.cpp(355): error: more than one operator "==" matches these operands:
            built-in operator "pointer == pointer"
            function "Rcpp::operator==(Rcpp::Na_Proxy, SEXP)"
            operand types are: Rcpp::AttributeProxyPolicy<Rcpp::RObject_Impl<Rcpp::PreserveStorage>>::AttributeProxy == SEXP
          if (col.attr("class") == R_NilValue) {
                                ^
```

So still not working. I hope this helps to pin down the issue here. Let me know if you need further tests. BTW, the boost config library is quite reliable to detect these compiler quirks.

--- a/inst/include/Rcpp/algorithm.h
+++ b/inst/include/Rcpp/algorithm.h
@@ -23,9 +23,11 @@
 #define Rcpp__Algorithm_h
 
 #if __cplusplus >= 201103L || __INTEL_CXX11_MODE__ == 1
-#    define RCPP_CONSTEXPR constexpr
+#    define RCPP_CONSTEXPR_FUNC constexpr
+#    define RCPP_CONSTEXPR_VAR constexpr
 #else
-#    define RCPP_CONSTEXPR const
+#    define RCPP_CONSTEXPR_FUNC
+#    define RCPP_CONSTEXPR_VAR const
 #endif
 
 namespace Rcpp {
@@ -160,41 +162,34 @@ namespace helpers {
 
     template< typename T >
     struct rtype_helper {
-        /*
-	typedef T type;
-        static RCPP_CONSTEXPR int RTYPE = REALSXP;
-        static inline double NA() { return NA_REAL; }
-        static inline RCPP_CONSTEXPR double ZERO() { return 0.0; }
-        static inline RCPP_CONSTEXPR double ONE() { return 1.0; }
-        */
     };
 
     template<>
     struct rtype_helper< double > {
         typedef double type;
-        static RCPP_CONSTEXPR int RTYPE = REALSXP;
+        static RCPP_CONSTEXPR_VAR int RTYPE = REALSXP;
         static inline double NA() { return NA_REAL; }
-        static inline RCPP_CONSTEXPR double ZERO() { return 0.0; }
-        static inline RCPP_CONSTEXPR double ONE() { return 1.0; }
+        static inline RCPP_CONSTEXPR_FUNC double ZERO() { return 0.0; }
+        static inline RCPP_CONSTEXPR_FUNC double ONE() { return 1.0; }
     };
 
     template<>
     struct rtype_helper< int > {
         typedef int type;
-        static RCPP_CONSTEXPR int RTYPE = INTSXP;
+        static RCPP_CONSTEXPR_VAR int RTYPE = INTSXP;
         static inline int NA() { return NA_INTEGER; }
-        static inline RCPP_CONSTEXPR int ZERO() { return 0; }
-        static inline RCPP_CONSTEXPR int ONE() { return 1; }
+        static inline RCPP_CONSTEXPR_FUNC int ZERO() { return 0; }
+        static inline RCPP_CONSTEXPR_FUNC int ONE() { return 1; }
     };
 
     template< typename T >
     struct rtype {
         typedef typename rtype_helper< typename ctype< T >::type >::type type;
         typedef rtype_helper< typename ctype< T >::type > helper_type;
-        static RCPP_CONSTEXPR int RTYPE = helper_type::RTYPE;
+        static RCPP_CONSTEXPR_VAR int RTYPE = helper_type::RTYPE;
         static inline T NA() { return helper_type::NA(); }
-        static inline RCPP_CONSTEXPR T ZERO() { return helper_type::ZERO(); }
-        static inline RCPP_CONSTEXPR T ONE() { return helper_type::ONE(); }
+        static inline RCPP_CONSTEXPR_FUNC T ZERO() { return helper_type::ZERO(); }
+        static inline RCPP_CONSTEXPR_FUNC T ONE() { return helper_type::ONE(); }
     };
 
     struct log {
@@ -488,6 +483,7 @@ void sqrt(InputIterator begin, InputIterator end, OutputIterator out) {
 } // namespace algorithm
 } // namespace Rcpp
 
-#undef RCPP_CONSTEXPR
+#undef RCPP_CONSTEXPR_FUNC
+#undef RCPP_CONSTEXPR_VAR
 
 #endif

